server/auth/types/multiple/multi_auth.ts

```diff
@@ -0,0 +1,192 @@
+    const routes = new MultiAuthRoutes(
+      this.router,
+      this.config,
+      this.sessionStorageFactory,
+      this.securityClient,
+      this.coreSetup
+    );
+    routes.setupRoutes();
+
+    for (let i = 0; i < this.authTypes.length; i++) {
+      switch (this.authTypes[i].toLowerCase()) {
+        case AuthType.BASIC: {
+          this.BasicAuth = new BasicAuthentication(
+            this.config,
+            this.sessionStorageFactory,
+            this.router,
+            this.esClient,
+            this.coreSetup,
+            this.logger
+          );
+          this.authHandlers.set(AuthType.BASIC, this.BasicAuth);
+          break;
+        }
+        case AuthType.OPEN_ID: {
+          this.OidcAuth = new OpenIdAuthentication(
+            this.config,
+            this.sessionStorageFactory,
+            this.router,
+            this.esClient,
+            this.coreSetup,
+            this.logger
+          );
+          this.authHandlers.set(AuthType.OPEN_ID, this.OidcAuth);
+          break;
+        }
+        case AuthType.SAML: {
+          this.SamlAuth = new SamlAuthentication(
+            this.config,
+            this.sessionStorageFactory,
+            this.router,
+            this.esClient,
+            this.coreSetup,
+            this.logger
+          );
+          this.authHandlers.set(AuthType.SAML, this.SamlAuth);
+          break;
+        }
+        default: {
+          throw new Error(`Unsupported authentication type: ${this.authTypes[i]}`);
+        }
+      }
+    }
+  }
+
+  // override functions inherited from AuthenticationType
+  async requestIncludesAuthInfo(
+    request: OpenSearchDashboardsRequest<unknown, unknown, unknown, any>
+  ): Promise<boolean> {
+    const cookie = await this.sessionStorageFactory.asScoped(request).get();
+    const reqAuthType = cookie?.authType?.toLowerCase();
+
+    if (reqAuthType && this.authHandlers.has(reqAuthType)) {
+      return this.authHandlers.get(reqAuthType)!.requestIncludesAuthInfo(request);
+    } else {
+      return false;
+    }
+  }
+
+  async getAdditionalAuthHeader(
+    request: OpenSearchDashboardsRequest<unknown, unknown, unknown, any>
+  ): Promise<any> {
+    const cookie = await this.sessionStorageFactory.asScoped(request).get();
+    const reqAuthType = cookie?.authType?.toLowerCase();
+
+    if (reqAuthType && this.authHandlers.has(reqAuthType)) {
+      return this.authHandlers.get(reqAuthType)!.getAdditionalAuthHeader(request);
+    } else {
+      return {};
+    }
+  }
+
+  async getCookie(
+    request: OpenSearchDashboardsRequest,
+    authInfo: any
+  ): Promise<SecuritySessionCookie> {
+    const cookie = await this.sessionStorageFactory.asScoped(request).get();
+    const reqAuthType = cookie?.authType?.toLowerCase();
+
+    if (reqAuthType && this.authHandlers.has(reqAuthType)) {
+      return this.authHandlers.get(reqAuthType)!.getCookie(request, authInfo);
+    } else {
+      return {};
+    }
+  }
```

Comment:
similar to `requestIncludesAuthInfo` the there might be no cookie when a REST client calling OSD API directly, and this code will fail

Reply:
Same login as above.

Reply:
you won't know which auth type the client is using based on request and auth info, in case of you have multiple auth types.

Reply:
Addressed.